src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMNavigator.java

```diff
@@ -163,12 +207,195 @@ public String getApiUri() {
+                }
+            }
+
+            if(apiKey == null) {
+                HttpGet getAuths = new HttpGet(apiBaseUrl + "/authorizations");
+                JSONArray auths = JSONArray.fromObject(getResponse(new UsernamePasswordCredentialsImpl(null, null, null, username, password), getAuths));
+                for (int i = 0; i < auths.size(); i++) {
+                    JSONObject auth = auths.getJSONObject(i);
+                    if (authName.equals(auth.getString("note"))) {
+                        // already exists
+                        response.put("error", "Already exists");
+                        return respondJson(500, response.toString());
+                    }
+                }
+
+                HttpPost createAuthPost = new HttpPost(apiBaseUrl + "/authorizations");
+
+                JSONObject json = new JSONObject();
+                json.put("note", authName);
+
+                JSONArray scopes = new JSONArray();
+                scopes.add("repo");
+                scopes.add("public_repo");
+                scopes.add("admin:org_hook");
+                scopes.add("admin:repo_hook");
+                json.put("scopes", scopes);
+
+                createAuthPost.setEntity(new StringEntity(json.toString()));
+
+                JSONObject res = JSONObject.fromObject(getResponse(new UsernamePasswordCredentialsImpl(null, null, null, username, password), createAuthPost));
+
+                if (res.containsKey("errors")) {
+                    JSONArray errors = res.getJSONArray("errors");
+                    JSONObject err = errors.getJSONObject(0);
+                    String code = err.getString("code");
+                    response.put("error", code);
+                }
+                else {
+                    apiKey = res.getString("token");
+                }
+            }
+
+            if(apiKey != null) {
+                // Would use the GitHubTokenCredentialsCreator, but the credential descriptions are not good
+                // and it creates incompatible credential types
+                StandardCredentials creds = createCredentials(apiUri, username, apiKey, uuid.substring(uuid.length()-4));
+
+                response.put("id", creds.getId());
+
+                return respondJson(200, response.toString());
             }
+        } catch(final Exception e) {
+            response.put("error", e.getMessage());
+        }
+
+        return respondJson(500, response.toString());
+    }
+
+    /**
+     * Create a github usesrname/password credential with the apiKey
+     */
+    private StandardCredentials createCredentials(String apiUri, String username, String apiKey, String identifier) {
+        String description = String.format("%s GitHub auth token *%s", username, identifier);
+        final UsernamePasswordCredentialsImpl creds = new UsernamePasswordCredentialsImpl(
+                CredentialsScope.GLOBAL,
+                UUID.randomUUID().toString(),
```

Comment:
Shouldn't we allow users to define the credentials.id so that it is predictable and can be reused in Jenkins files or pipeline scripts?